File: sdk/ts/src/submissions.ts
import type { VelocityClient } from './client.js'

export type SubmissionKind = 'matching_engine' | 'market_maker' | 'pcap_replay'

export type Submission = {
  id: string
  team: string
  display: string
  kind: SubmissionKind
  status: string
  created_at_ns: number
}

export type CreateSubmissionRequest = {
  team: string
  display: string
  kind: SubmissionKind
  source?: Blob | ArrayBuffer | Uint8Array
}

export class SubmissionsResource {
  constructor(private readonly client: VelocityClient) {}

  async create(req: CreateSubmissionRequest): Promise<Submission> {
    const resp = await this.client.request<{ submission: Submission; upload_url?: string }>(
      'POST', '/v1/submissions',
      { team: req.team, display: req.display, kind: req.kind },
    )
    if (req.source && resp.upload_url) {
      const body =
        req.source instanceof Blob ? req.source :
        // Copy into a fresh ArrayBuffer-backed view: a bare `Uint8Array`
        // is `Uint8Array<ArrayBufferLike>` (TS 5.7+), which `BlobPart`
        // rejects because the backing buffer could be a SharedArrayBuffer.
        req.source instanceof Uint8Array ? new Blob([new Uint8Array(req.source)]) :
        new Blob([req.source])
      // Direct PUT to the signed URL — no auth header because the
      // signature is in the URL query params.
      const res = await fetch(resp.upload_url, { method: 'PUT', body })
      if (!res.ok) throw new Error(`source upload failed: HTTP ${res.status}`)
      await this.client.request<void>('POST',
        `/v1/submissions/${encodeURIComponent(resp.submission.id)}/uploaded`,
        { id: resp.submission.id },
      )
    }
    return resp.submission
  }

  async get(id: string): Promise<Submission> {
    const resp = await this.client.request<{ submission: Submission }>(
      'GET', `/v1/submissions/${encodeURIComponent(id)}`,
    )
    return resp.submission
  }

  async list(limit = 20): Promise<Submission[]> {
    const resp = await this.client.request<{ submissions: Submission[] }>(
      'GET', `/v1/submissions?limit=${limit}`,
    )
    return resp.submissions
  }
}

File: sdk/go/submissions.go
package velocity

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
)

type SubmissionKind string

const (
	SubmissionKindMatchingEngine SubmissionKind = "matching_engine"
	SubmissionKindMarketMaker    SubmissionKind = "market_maker"
	SubmissionKindPcapReplay     SubmissionKind = "pcap_replay"
)

type Submission struct {
	ID        string         `json:"id"`
	Team      string         `json:"team"`
	Display   string         `json:"display"`
	Kind      SubmissionKind `json:"kind"`
	Status    string         `json:"status"`
	CreatedAt int64          `json:"created_at_ns"`
}

type CreateSubmissionRequest struct {
	Team        string         `json:"team"`
	Display     string         `json:"display"`
	Kind        SubmissionKind `json:"kind"`
	SourceTarGz io.Reader      `json:"-"`
}

// SubmissionsResource bundles all /v1/submissions endpoints.
type SubmissionsResource struct{ c *Client }

// Create uploads source and registers a new submission. The artefact
// upload is a separate hop: the gateway first allocates an ID + signed
// upload URL, the SDK PUTs the source to MinIO, then the gateway is
// notified. We hide all that here so the caller just hands us a reader.
func (r *SubmissionsResource) Create(ctx context.Context, req *CreateSubmissionRequest) (*Submission, error) {
	body, err := json.Marshal(map[string]any{
		"team":    req.Team,
		"display": req.Display,
		"kind":    req.Kind,
	})
	if err != nil {
		return nil, err
	}
	var resp struct {
		Submission Submission `json:"submission"`
		UploadURL  string     `json:"upload_url"`
	}
	if err := r.c.do(ctx, "POST", "/v1/submissions",
		bytes.NewReader(body), &resp); err != nil {
		return nil, err
	}
	if req.SourceTarGz != nil && resp.UploadURL != "" {
		if err := r.upload(ctx, resp.UploadURL, req.SourceTarGz); err != nil {
			return nil, err
		}
		// Notify the gateway that upload completed. The route is a
		// no-op when the gateway is configured against an inline
		// MinIO-event hook, but explicitly notifying makes the SDK
		// portable across deployments.
		notifyBody, _ := json.Marshal(map[string]any{"id": resp.Submission.ID})
		if err := r.c.do(ctx, "POST",
			"/v1/submissions/"+resp.Submission.ID+"/uploaded",
			bytes.NewReader(notifyBody), nil); err != nil {
			return nil, err
		}
	}
	return &resp.Submission, nil
}

// Get fetches a submission by id.
func (r *SubmissionsResource) Get(ctx context.Context, id string) (*Submission, error) {
	var resp struct {
		Submission Submission `json:"submission"`
	}
	if err := r.c.do(ctx, "GET", "/v1/submissions/"+id, nil, &resp); err != nil {
		return nil, err
	}
	return &resp.Submission, nil
}

// List returns up to `limit` recent submissions for the current tenant.
func (r *SubmissionsResource) List(ctx context.Context, limit int) ([]Submission, error) {
	var resp struct {
		Submissions []Submission `json:"submissions"`
	}
	if err := r.c.do(ctx, "GET",
		"/v1/submissions?limit="+intToString(limit), nil, &resp); err != nil {
		return nil, err
	}
	return resp.Submissions, nil
}

func (r *SubmissionsResource) upload(ctx context.Context, url string, src io.Reader) error {
	// Direct PUT to the signed URL. We deliberately skip the SDK's
	// retry helper here — MinIO signed URLs are time-limited and a
	// long retry tail does more harm than good.
	body, err := io.ReadAll(src)
	if err != nil {
		return err
	}
	req, err := newPlainRequest(ctx, "PUT", url, body)
	if err != nil {
		return err
	}
	resp, err := r.c.http.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return &APIError{StatusCode: resp.StatusCode, Message: "upload failed"}
	}
	return nil
}
